lib/spack/llnl/util/filesystem.py

```diff
@@ -242,6 +248,100 @@ def compressed_mime_type(filename):
     return type, subtype
 
 
+@contextmanager
+def stream_in_place(stream):
+    """Ensures that stream is returned to starting position
+    once context manager exits"""
+    current_stream_pos = stream.tell()
+    yield stream
+    stream.seek(current_stream_pos)
+
+
+def compose_bin(bin):
+    """Return numeric value of concatenated adjacent bytes"""
+    p = 0
+    for count, c in enumerate(bin[::-1]):
+        root_c = c
+        for _ in range(count*2):
+            root_c = root_c << 8
+        p = p + root_c
+    return p
+
+
+class WinPE:
+    """Base class for Windows Portable Executable (PE)
+    file type characteristic definitions"""
+    signature = b"PE\x00\x00"
+    flag = 0
+    offset = 0x3c
+    type = "application"
+
+    @classmethod
+    def validate(cls, stream):
+        """Validate PE type (dll vs exe)"""
+        with stream_in_place(stream) as s:
+            # PE signature is offset by whatever value is at the
+            # address 0x3c in the binary file
+            sig_offset = s.read(0x3c+1)[0x3c]
+            sig = s.read(sig_offset-1-0x32+4)[-4:]
+            # Signature does not match general PE format
+            # short circuit as resolving subtype doesn't matter
+            if not sig == cls.signature:
+                return False
+            # We know the file is a win PE format
+            # now determine dll vs exe
+            # flag to determine file characteristics is
+            # offset by end of signature by 18 bytes
+            characteristics = s.read(20)[-2:]
+            hex_char = compose_bin(characteristics)
+            # Check for appropriate characteristics flag
+            return (hex_char & cls.flag) == cls.flag
+
+
+class WinExe(WinPE):
+    """Class defining the characteristics of a Windows EXE
+    file"""
+    sub_type = "x-executable"
+    flag = 0x0002
+
+
+class WinDll(WinPE):
+    """Class defining the characteristics of a Windows DLL
+    file"""
+    sub_type = "x-sharedlib"
+    flag = 0x2000
+
+class WinCoff:
+    """Class defining characteristics of a Windows
+    Common Object File Format (COFF) file
+
+    This is both static libraries and dll import libs
+    """
+    signature = b"!<arch>\n"
+    sub_type = "static-library"
+    type = "archive"
+    offset = 0
+
+    @classmethod
+    def validate(cls, stream):
+        """Return true if stream's first bytes are the first
+        few bytes of a valid COFF file"""
+        with stream_in_place(stream) as s:
+            read_sig = s.read(cls.offset + len(cls.signature))
+        return read_sig == cls.signature
+
+
+SUPPORTED_WIN_TYPES = [WinCoff, WinDll, WinExe]
+
+
+def get_win_pe_coff_type(file):
```

Comment:
Since this also returns types for `WinDll` and `WinExe`, those are also COFF types? The naming and organization of attributes is a bit confusing (e.g.`WinCoff` is always a `static-library`).

Reply:
`pe_coff`: `PE` - portable executable, so in addition to the COFF files (coff archives/import libs and .obj files) this will also correctly identify binaries, like dlls and exes.